### PR TITLE
Allow `device_mesh` have multiple dim 

### PR DESCRIPTION
**Pull request by S1ro1**
Experimental PR for me to just make stuff work on the accelerate side, will follow up with this after.

### CODE REVIEW

**Review by winglian — APPROVED** [review]
This looks like all the needed changes to me! (plus https://github.com/huggingface/transformers/pull/39562)

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -4317,10 +4318,11 @@ def from_pretrained(
             if device_mesh is None and tp_plan is not None:
                 tp_plan, device_map, device_mesh = initialize_tensor_parallelism(tp_plan, tp_size=None)
             else:
-                # TODO: make device_mesh support multiple dimensions
-                if device_mesh.ndim == 1:
-                    raise ValueError("device_mesh must be 1 dimensional and will be used for TP")
-                device_map = torch.device(device_mesh.device_type, int(os.environ["LOCAL_RANK"]))
+                assert device_mesh is not None and "tp" in device_mesh.mesh_dim_names, (
+                    "device_mesh must contain a 'tp' dimension"
+                )
```
Comment (winglian) [inline src/transformers/modeling_utils.py]: can simplify some of this logic
```
diff --git a/src/transformers/modeling_utils.py b/src/transformers/modeling_utils.py
index f383ba1ad3..262f8efb97 100644
--- a/src/transformers/modeling_utils.py
+++ b/src/transformers/modeling_utils.py
@@ -4682,12 +4682,11 @@ class PreTrainedModel(nn.Module, EmbeddingAccessMixin, ModuleUtilsMixin, PushToH
         # We need to correctly dispatch the model on the current process device. The easiest way for this is to use a simple
         # `device_map` pointing to the correct device
         if tp_plan is not None:
-            if device_mesh is None and tp_plan is not None:
+            if device_mesh is None:
                 tp_plan, device_map, device_mesh = initialize_tensor_parallelism(tp_plan, tp_size=None)
             else:
-                assert device_mesh is not None and "tp" in device_mesh.mesh_dim_names, (
-                    "device_mesh must contain a 'tp' dimension"
-                )
+                if "tp" not in device_mesh.mesh_dim_names:
+                    raise ValueError("device_mesh must contain a 'tp' dimension")
                 device_mesh = device_mesh["tp"]
                 tp_size = device_mesh["tp"].size()
```

**Review by SunMarc — APPROVED** [review]
SGTM !
